### PR TITLE
remove finalizers from the CR if operator get deleted first

### PR DESCRIPTION
**Pull request by slayer321**
##### Checklist

- [x] Tested in playground or other setup
- [ ] Screenshot (Grafana) from playground added to PR for 15+ minute run
- [ ] Documentation is changed or added
- [ ] Tests and/or benchmarks are included
- [ ] Breaking changes

### Summary by CodeRabbit

**Release Notes**

- **Refactor**: Renamed methods `updateAgent` to `UpdateAgent` and `updateController` to `UpdateController` for better code readability.
- **New Feature**: Introduced functions `RemoveFinalizerFromAgentCR` and `RemoveFinalizerFromControllerCR` to handle operator deletion more gracefully by removing finalizers from the Agent and Controller Custom Resources respectively.
- **Refactor**: Added a new function `setupContext` for setting up application context and signal handling, enhancing the robustness of the application.

> 🎉 Here's to the code that we refine,  
> With each pull request, it does shine.  
> Finalizers removed, signals caught,  
> Graceful shutdowns as we sought.  
> In our README, no badge out of line,  
> Oh, how beautifully does our codebase align! 🎉

### CODE REVIEW

File: operator/main.go [hunk 1]
```diff
@@ -240,3 +246,38 @@ func main() {
...
+		cancel()
+	}()
+	return ctx
+}
+
+func removeFinalizer(ctx context.Context, reconciler *agent.AgentReconciler) {
+	// Remove the finalizer from the agent
+	agent := &agentv1alpha1.Agent{}
+	err := reconciler.Client.Get(ctx, types.NamespacedName{
+		Name:      os.Getenv("AGENT_CR_NAME"),
```
Comment (hdkshingala) [inline operator/main.go:268]: This logic is restricted to the Helm chart based installation only, which is not correct.

What if someone installs operator separately and then the CR?

File: operator/controllers/agent/reconciler.go [hunk 2]
```diff
@@ -120,7 +120,7 @@ func (r *AgentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 			r.deleteResources(ctx, logger, instance.DeepCopy())
 
 			controllerutil.RemoveFinalizer(instance, controllers.FinalizerName)
-			if err = r.updateAgent(ctx, instance); err != nil && !errors.IsNotFound(err) {
+			if err = r.UpdateAgent(ctx, instance); err != nil && !errors.IsNotFound(err) {
 				return ctrl.Result{}, err
 			}
 		}
```
[the remark this answers is not in the document]
Comment (slayer321) [reply operator/controllers/agent/reconciler.go:126]: @openai do you know how can I use it still keeping it private in main.go file ?

File: operator/controllers/agent/reconciler.go [hunk 3]
```diff
@@ -846,6 +846,31 @@ func (r *AgentReconciler) reconcileNamespacedResources(ctx context.Context, log
 	return nil
 }
 
+// RemoveFinalizerFromAgentCR removes the finalizer from the Agent CR when the operator is getting deleted.
+func (r *AgentReconciler) RemoveFinalizerFromAgentCR(ctx context.Context, setupLog logr.Logger) {
+	agentList := &agentv1alpha1.AgentList{}
+	err := r.Client.List(ctx, agentList)
+	if err != nil {
+		setupLog.Error(err, "Error while getting the agent")
```
Comment (hdkshingala) [inline operator/controllers/agent/reconciler.go:854]: We have to return from here as there is no meaning to continue execution after this.

File: operator/controllers/agent/reconciler.go [hunk 4]
```diff
@@ -846,6 +846,31 @@ func (r *AgentReconciler) reconcileNamespacedResources(ctx context.Context, log
...
+	}
+	if agentList.Items != nil && len(agentList.Items) != 0 {
+		for _, agentCR := range agentList.Items {
+			agentCR := agentCR
+			if controllerutil.ContainsFinalizer(&agentCR, controllers.FinalizerName) {
+				setupLog.Info("Operator is getting deleted. Removing finalizer from the Agent CR")
+				controllerutil.RemoveFinalizer(&agentCR, controllers.FinalizerName)
+				if err = r.updateAgent(ctx, &agentCR); err != nil && !errors.IsNotFound(err) {
+					if err.Error() == "Unauthorized" {
+						setupLog.Error(err, "Unauthorized to remove Finalizer from the agent serviceaccount might be deleted")
```
Comment (hdkshingala) [inline operator/controllers/agent/reconciler.go:864]: Same as above

File: operator/controllers/agent/reconciler.go [hunk 5]
```diff
@@ -846,6 +846,31 @@ func (r *AgentReconciler) reconcileNamespacedResources(ctx context.Context, log
...
+					} else {
+						setupLog.Error(err, "Error while removing Finalizer from the agent")
```
Comment (hdkshingala) [inline operator/controllers/agent/reconciler.go:866]: Same as above

File: operator/controllers/controller/reconciler.go [hunk 6]
```diff
@@ -734,6 +734,32 @@ func (r *ControllerReconciler) reconcileSecret(ctx context.Context, instance *co
 	return nil
 }
 
+// RemoveFinalizerFromControllerCR removes the finalizer from the controller CR when the operator is getting deleted.
+func (r *ControllerReconciler) RemoveFinalizerFromControllerCR(ctx context.Context, setupLog logr.Logger) {
+	controllerList := &controllerv1alpha1.ControllerList{}
+
+	err := r.Client.List(ctx, controllerList)
+	if err != nil {
+		setupLog.Error(err, "Error while getting the controller")
```
Comment (hdkshingala) [inline operator/controllers/controller/reconciler.go:743]: Same as above

File: operator/controllers/controller/reconciler.go [hunk 7]
```diff
@@ -734,6 +734,32 @@ func (r *ControllerReconciler) reconcileSecret(ctx context.Context, instance *co
...
+	}
+	if controllerList.Items != nil && len(controllerList.Items) != 0 {
+		for _, controllerCR := range controllerList.Items {
+			controllerCR := controllerCR
+			if controllerutil.ContainsFinalizer(&controllerCR, controllers.FinalizerName) {
+				setupLog.Info("Operator is getting deleted. Removing finalizer from the Controller CR")
+				controllerutil.RemoveFinalizer(&controllerCR, controllers.FinalizerName)
+				if err = r.updateController(ctx, &controllerCR); err != nil && !errors.IsNotFound(err) {
+					if err.Error() == "Unauthorized" {
+						setupLog.Error(err, "Unauthorized to remove Finalizer from the controller serviceaccount might be deleted")
```
Comment (hdkshingala) [inline operator/controllers/controller/reconciler.go:753]: Same as above

File: operator/controllers/controller/reconciler.go [hunk 8]
```diff
@@ -734,6 +734,32 @@ func (r *ControllerReconciler) reconcileSecret(ctx context.Context, instance *co
...
+					} else {
+						setupLog.Error(err, "Error while removing Finalizer from the controller")
```
Comment (hdkshingala) [inline operator/controllers/controller/reconciler.go:755]: Same as above

**Review by hdkshingala — CHANGES_REQUESTED** [review]

File: operator/controllers/agent/reconciler.go [hunk 9]
```diff
@@ -846,6 +846,34 @@ func (r *AgentReconciler) reconcileNamespacedResources(ctx context.Context, log
...
+		return
+	}
+	if agentList.Items != nil && len(agentList.Items) != 0 {
+		for _, agentCR := range agentList.Items {
+			agentCR := agentCR
+			if controllerutil.ContainsFinalizer(&agentCR, controllers.FinalizerName) {
+				setupLog.Info("Operator is getting deleted. Removing finalizer from the Agent CR")
+				controllerutil.RemoveFinalizer(&agentCR, controllers.FinalizerName)
+				if err = r.updateAgent(ctx, &agentCR); err != nil && !errors.IsNotFound(err) {
+					if err.Error() == "Unauthorized" {
```
Comment (hdkshingala) [inline operator/controllers/agent/reconciler.go:864]: I think there is an built-in option for check this in `errors` library. Please verify that

**Review by hdkshingala — APPROVED** [review]

**Pull request merged**